logins-sql/src/db.rs

```diff
@@ -0,0 +1,789 @@
+        Ok(logins)
+    }
+
+    pub fn open(path: impl AsRef<Path>, encryption_key: Option<&str>) -> Result<Self> {
+        Ok(Self::with_connection(Connection::open(path)?, encryption_key)?)
+    }
+
+    pub fn open_in_memory(encryption_key: Option<&str>) -> Result<Self> {
+        Ok(Self::with_connection(Connection::open_in_memory()?, encryption_key)?)
+    }
+
+    pub fn execute_all(&self, stmts: &[&str]) -> Result<()> {
+        for sql in stmts {
+            self.execute(sql, &[])?;
+        }
+        Ok(())
+    }
+
+    #[inline]
+    pub fn execute(&self, stmt: &str, params: &[(&str, &ToSql)]) -> Result<usize> {
+        Ok(self.do_exec(stmt, params, false)?)
+    }
+
+    #[inline]
+    pub fn execute_cached(&self, stmt: &str, params: &[(&str, &ToSql)]) -> Result<usize> {
+        Ok(self.do_exec(stmt, params, true)?)
+    }
+
+    fn do_exec(&self, sql: &str, params: &[(&str, &ToSql)], cache: bool) -> Result<usize> {
+        let res = if cache {
+            self.db.prepare_cached(sql)
+                   .and_then(|mut s| s.execute_named(params))
+        } else {
+            self.db.execute_named(sql, params)
+        };
+        if let Err(e) = &res {
+            warn!("Error running SQL {}. Statement: {:?}", e, sql);
+        }
+        Ok(res?)
+    }
+
+    pub fn query_one<T: FromSql>(&self, sql: &str) -> Result<T> {
+        let res: T = self.db.query_row(sql, &[], |row| row.get(0))?;
+        Ok(res)
+    }
+
+    // The type returned by prepare/prepare_cached are different, but must live
+    // to the end of the function, so (AFAICT) it's difficult/impossible to
+    // remove the duplication between query_row/query_row_cached.
+
+    pub fn query_row<T>(&self, sql: &str, args: &[(&str, &ToSql)], f: impl FnOnce(&Row) -> Result<T>) -> Result<Option<T>> {
+        let mut stmt = self.db.prepare(sql)?;
+        let res = stmt.query_named(args);
+        if let Err(e) = &res {
+            warn!("Error executing query: {}. Query: {}", e, sql);
+        }
+        let mut rows = res?;
+        match rows.next() {
+            Some(result) => Ok(Some(f(&result?)?)),
+            None => Ok(None),
+        }
+    }
+
+    pub fn query_row_cached<T>(&self, sql: &str, args: &[(&str, &ToSql)], f: impl FnOnce(&Row) -> Result<T>) -> Result<Option<T>> {
+        let mut stmt = self.db.prepare_cached(sql)?;
+        let res = stmt.query_named(args);
+        if let Err(e) = &res {
+            warn!("Error executing query: {}. Query: {}", e, sql);
+        }
+        let mut rows = res?;
+        match rows.next() {
+            Some(result) => Ok(Some(f(&result?)?)),
+            None => Ok(None),
+        }
+    }
+}
+
+// login specific stuff.
+
+impl LoginDb {
+
+    fn mark_as_synchronized(&mut self, guids: &[&str], ts: ServerTimestamp) -> Result<()> {
+        util::each_chunk(guids, self.max_var_count, |chunk, _| {
+            self.db.execute(
+                &format!("DELETE FROM {mirror_table} WHERE guid IN ({vars})",
```

Comment:
Are we planning to do three-way merges?

Reply:
We should already be doing three way merges!